internal/controller/imagerepository_controller.go

```diff
@@ -877,3 +879,226 @@ func (r *ImageRepositoryReconciler) ImageRepositoryForSameUrlExists(ctx context.
 func getComponentSaName(componentName string) string {
 	return fmt.Sprintf("%s%s", componentSaNamePrefix, componentName)
 }
+
+// addPullSecretToApplicationPullSecret updates the application pull secret when new image repository pull secret is created
+// or when an existing one is updated.
+func (r *ImageRepositoryReconciler) addPullSecretToApplicationPullSecret(ctx context.Context, applicationName string, namespace string, pullSecretName string) error {
+	log := ctrllog.FromContext(ctx)
+
+	application := &appstudioredhatcomv1alpha1.Application{}
+	if err := r.Client.Get(ctx, types.NamespacedName{Name: applicationName, Namespace: namespace}, application); err != nil {
+		log.Error(err, "failed to get Application", "application", applicationName)
+		return err
+	}
+
+	applicationPullSecretName := getApplicationPullSecretName(applicationName)
+	applicationPullSecret := &corev1.Secret{}
+	if err := r.Client.Get(ctx, types.NamespacedName{Name: applicationPullSecretName, Namespace: namespace}, applicationPullSecret); err != nil {
+		log.Error(err, "failed to get application pull secret", "secretName", applicationPullSecretName)
+		return err
+	}
+
+	pullSecret := &corev1.Secret{}
+	if err := r.Client.Get(ctx, types.NamespacedName{Name: pullSecretName, Namespace: namespace}, pullSecret); err != nil {
+		log.Error(err, "failed to get individual pull secret", "secretName", pullSecretName)
+		return err
+	}
+	if pullSecret.Type != corev1.SecretTypeDockerConfigJson {
+		log.Info("Skipping secret due to unexpected type", "secretName", pullSecretName, "type", pullSecret.Type)
+		return nil
+	}
+
+	dockerConfigDataBytes, ok := pullSecret.Data[corev1.DockerConfigJsonKey]
+	if !ok {
+		log.Info("Missing .dockerconfigjson key", "secretName", pullSecretName)
+		return nil
+	}
+
+	var newAuths dockerConfigJson
+	if err := json.Unmarshal(dockerConfigDataBytes, &newAuths); err != nil {
+		log.Error(err, "failed to unmarshal .dockerconfigjson", "secretName", pullSecretName)
+		return err
+	}
+
+	var existingAuths dockerConfigJson
+	if err := json.Unmarshal(applicationPullSecret.Data[corev1.DockerConfigJsonKey], &existingAuths); err != nil {
+		log.Error(err, "failed to unmarshal existing .dockerconfigjson", "secretName", applicationPullSecretName)
+		return err
+	}
+
+	changed := false
+	if existingAuths.Auths == nil {
+		existingAuths.Auths = make(map[string]struct {
+			Auth string `json:"auth"`
+		})
+	}
+
+	// If there are multiple pullsecrets for the same registry,
+	// they are overriden and we keep only the latest
+	for reg, entry := range newAuths.Auths {
+		if existingAuths.Auths[reg].Auth != entry.Auth {
+			existingAuths.Auths[reg] = entry
```

Comment:
I think we should just check if reg matches and isn't empty
and only then add it

no point of updating it to new pull secret from new IR
since old will still work, and removal of old one will add there different existing

we safe 1 update